include/exadg/operators/operator_base.cpp

```diff
@@ -2153,6 +2157,175 @@ OperatorBase<dim, Number, n_components>::evaluate_face_integrals() const
+
+    dealii::SparseMatrixTools::restrict_to_full_matrices<SparseMatrix,
+                                                         dealii::DynamicSparsityPattern,
+                                                         Number>(tmp_matrix,
+                                                                 dsp,
+                                                                 dof_indices_all_cells,
+                                                                 overlapped_cell_matrices);
+
+    // factorize and store cell matrices
+    matrices.resize(n_cells, LAPACKMatrix(dofs_per_cell));
+    for(unsigned int cell_batch_index = 0; cell_batch_index < matrix_free->n_cell_batches();
+        ++cell_batch_index)
+    {
+      unsigned int const n_filled_lanes =
+        matrix_free->n_active_entries_per_cell_batch(cell_batch_index);
+
+      for(unsigned int v = 0; v < n_filled_lanes; ++v)
+      {
+        // get overlapped cell matrix
+        auto const & overlapped_cell_matrix =
+          overlapped_cell_matrices[cell_batch_index * vectorization_length + v];
+
+        // store the cell matrix and renumber lexicographic
+        auto & lapack_matrix = matrices[cell_batch_index * vectorization_length + v];
+
+        auto const & lex = matrix_free->get_shape_info().lexicographic_numbering;
+        for(unsigned int i = 0; i < dofs_per_cell; i++)
+          for(unsigned int j = 0; j < dofs_per_cell; j++)
+            lapack_matrix.set(i, j, overlapped_cell_matrix[lex[i]][lex[j]]);
+
+        // factorize the cell matrix
+        lapack_matrix.compute_lu_factorization();
+      }
+    }
+  }
+}
+
+template<int dim, typename Number, int n_components>
+void
+OperatorBase<dim, Number, n_components>::apply_inverse_additive_schwarz_matrices(
+  VectorType &       dst,
+  VectorType const & src) const
+{
+  if(is_dg)
+    apply_inverse_block_diagonal(dst, src);
+  else
+  {
+    src.update_ghost_values();
+
+    matrix_free->template cell_loop<VectorType, VectorType>(
+      [&](auto const & matrix_free, auto & dst, auto const & src, auto const & cell_range) {
+        auto const dofs_per_cell = matrix_free.get_dofs_per_cell(this->data.dof_index);
+
+        IntegratorCell integrator =
+          IntegratorCell(matrix_free, this->data.dof_index, this->data.quad_index);
+
+        dealii::Vector<Number>                                 local_vector(dofs_per_cell);
+        dealii::AlignedVector<dealii::VectorizedArray<Number>> local_weights_vector(dofs_per_cell);
+        for(unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+        {
+          this->reinit_cell(integrator, cell);
+
+          integrator.read_dof_values(weights);
+          for(unsigned int i = 0; i < dofs_per_cell; ++i)
+            local_weights_vector[i] = integrator.begin_dof_values()[i];
+
+          integrator.read_dof_values(src);
+
+          for(unsigned int v = 0; v < vectorization_length; ++v)
```

Comment:
The function above uses `n_filled_lanes` and not `vectorization_length`. Could this become a problem?

Reply:
Put differently, what is the difference between `cell` used here and `cell_batch_index` used above?

Reply:
> The function above uses n_filled_lanes and not vectorization_length. Could this become a problem?

It should not, `FEEvaluation::read_dof_values` and `FEEvaluation::distribute_local_to_global` ignore the lanes which are not filled.

Reply:
> Put differently, what is the difference between cell used here and cell_batch_index used above?

That's a question of philosophy, we usually call the cell batch number `cell` in matrix-free loops. I like being explict here and would at least call it `cell_batch` or `cell_batch_index` like above.

Standard deal.II often writes `cell` for a `DofCellAccessor/CellAccessor` like in the function above. We should aim for a consistent naming, so for now `cell` for the cell batch index and `cell_v` for the accessor (to be consistent with current ExaDG)?

Reply:
But we still execute `matrix.solve()` for such non-filled lanes? If I understand it correctly, we did not compute / factorize such matrices before solving?

Reply:
Regarding the naming: or something with "cell_iterator" / "cell_pointer" for the accessor to be clear.

Reply:
This is consistent to current `OperatorBase` 2ecad0db2861d7461986128eec087cbce80133bc

Reply:
And we have to use `n_filled_lanes` here, you are right 02d181ae3d97050a6e2c8ce38a75a2885f4cfaf6

Not easy to trigger with global refinements.